table/transaction.go

```diff
@@ -121,6 +140,140 @@ func (t *Transaction) Append(rdr array.RecordReader, snapshotProps iceberg.Prope
+	var (
+		inputFiles     = make(map[string]bool)
+		outputFiles    = make(map[string]struct{})
+		numInputFiles  int
+		numOutputFiles int
+	)
+
+	for _, op := range ops {
+		if len(op.InputFiles) <= 1 {
+			return fmt.Errorf("merge operation must have at least 2 input files (%d)", len(op.InputFiles))
+		}
+
+		outputFiles[op.OutputFile] = struct{}{}
+		numOutputFiles++
+
+		for _, f := range op.InputFiles {
+			inputFiles[f] = false
+			numInputFiles++
+		}
+	}
+
+	if len(outputFiles) != numOutputFiles {
+		return errors.New("duplicate output files")
+	}
+
+	if len(inputFiles) != numInputFiles {
+		return errors.New("duplicate input files")
+	}
+
+	s := t.meta.currentSnapshot()
+
+	if s == nil {
+		return errors.New("merge operation requires an existing snapshot")
+	}
+
+	manifestFiles, err := s.Manifests(t.tbl.fs)
+	if err != nil {
+		return err
+	}
+
+	var (
+		existingManifestFiles  []iceberg.ManifestFile // manifest that don't contain any input files
+		existingDataFiles      []iceberg.DataFile     // existing data file entries in manifest that contain some input files
+		deletedManifestEntries []iceberg.ManifestEntry
+		deletedManifestFiles   []iceberg.ManifestFile
+	)
+
+	for _, manifestFile := range manifestFiles {
+		entries, err := manifestFile.FetchEntries(t.tbl.fs, false)
+		if err != nil {
+			return err
+		}
+
+		var (
+			isManifestFileTouched bool
+			untouchedDataFiles    []iceberg.DataFile
+		)
+
+		for _, entry := range entries {
+			entry.Status()
+
+			_, found := inputFiles[entry.DataFile().FilePath()]
+
+			if !found {
+				untouchedDataFiles = append(untouchedDataFiles, entry.DataFile())
+			} else {
+				inputFiles[entry.DataFile().FilePath()] = true
+				isManifestFileTouched = true
+
+				deletedManifestEntries = append(deletedManifestEntries, iceberg.NewManifestEntry(
+					iceberg.EntryStatusDELETED,
+					internal.ToPtr(entry.SnapshotID()),
+					internal.ToPtr(entry.SequenceNum()),
+					entry.FileSequenceNum(),
+					entry.DataFile(),
+				))
+			}
+		}
+
+		if !isManifestFileTouched {
+			existingManifestFiles = append(existingManifestFiles, manifestFile)
+		} else {
+			deletedManifestFiles = append(deletedManifestFiles, manifestFile)
+			existingDataFiles = append(existingDataFiles, untouchedDataFiles...)
+
+		}
+	}
+
+	for f, found := range inputFiles {
+		if !found {
+			return fmt.Errorf("input file %s not found in any manifest", f)
+		}
+	}
+
+	if t.meta.NameMapping() == nil {
+		nameMapping := t.meta.CurrentSchema().NameMapping()
+		mappingJson, err := json.Marshal(nameMapping)
+		if err != nil {
+			return err
+		}
+		err = t.SetProperties(iceberg.Properties{DefaultNameMappingKey: string(mappingJson)})
+		if err != nil {
+			return err
+		}
+	}
+
+	updater := t.updateSnapshot(snapshotProps).merge(existingManifestFiles, deletedManifestFiles, deletedManifestEntries)
+
+	outputDataFiles := parquetFilesToDataFiles(t.tbl.fs, t.meta, maps.Keys(outputFiles))
```

Comment:
why are we re-processing all of the files? We have manifest entries for them already, we don't need to re-read and process them to create new entries. We can just use the existing entries, right?